### ISSUE TITLE
B4 - Multi-Label Classification Auxiliary Loss

### ISSUE THREAD

**Issue by iremurhan** [iremurhan/tez_v2_clean#58]
## Motivation
 
Our ablation results show that loss-level (B1) and data-level (B2) interventions yield negligible or negative gains on retrieval. The bottleneck appears to be the **representation itself** — CLIP's CLS embedding doesn't encode which objects are present in the image.
 
İnci Hoca proposed injecting object-awareness into the ViT output by adding a **multi-label classification head** trained with an auxiliary loss. Since COCO already provides per-image object category annotations (80 classes), this requires no external model — just leveraging existing ground-truth labels.
 
### Architecture
 
```
CLIP ViT image encoder
        │
    CLS token embedding (d=768)
        │
        ├──────────────────────┐
        │                      │
  Contrastive head        Classification head
  (existing, projects      (NEW: Linear(768, 80))
   to shared space)              │
        │                   sigmoid per class
        │                      │
  L_contrastive           L_cls = BCE(pred, multi-hot label)
        │                      │
        └──────────┬───────────┘
                   │
         L_total = L_contrastive + λ · L_cls
```
 
### λ Schedule
 
- Start with `λ = 1.0` (classification signal is strong early on).
- Anneal during training (e.g., cosine decay or step-wise reduction).
- Rationale: early epochs → learn "what objects exist"; later epochs → focus on fine-grained contrastive alignment.
- First experiment: keep `λ = 1.0` fixed to isolate the effect. Second experiment: try annealing.
 
### 1. Ground-Truth Label Integration
 
COCO provides `instances_train2014.json` and `instances_val2014.json` with per-image object annotations. Each image has a list of annotations, each with a `category_id` (80 categories).
 
**Steps:**
1. Parse COCO annotations JSON and build a mapping: `image_id → multi-hot vector of shape (80,)`.
2. Pre-compute and cache this mapping (e.g., as a `.pt` or `.pkl` file) so it doesn't slow down the dataloader.
3. Modify the dataloader to return `(image, caption, label_vector)` instead of `(image, caption)`.
 
**Code sketch for label extraction:**
```python
from pycocotools.coco import COCO
 
coco = COCO("annotations/instances_train2014.json")
cat_ids = sorted(coco.getCatIds())  # 80 category IDs
cat_id_to_idx = {cat_id: idx for idx, cat_id in enumerate(cat_ids)}
 
label_map = {}  # image_id -> tensor of shape (80,)
for img_id in coco.getImgIds():
    ann_ids = coco.getAnnIds(imgIds=img_id)
    anns = coco.loadAnns(ann_ids)
    vec = torch.zeros(80)
    for ann in anns:
        vec[cat_id_to_idx[ann["category_id"]]] = 1.0
    label_map[img_id] = vec
```
 
> **Note:** Flickr30K does not have standard object category annotations, so this experiment is COCO-only. Flickr evaluation will use the same model but without the classification loss active.
 
### 2. Classification Head
 
- Single `nn.Linear(d_model, 80)` on top of the CLS embedding.
- No extra hidden layers — keep it minimal so the gradient signal goes directly into the ViT.
- Apply `sigmoid` (not softmax) since multiple classes can co-occur.
 
### 3. Loss Integration
 
```python
cls_logits = cls_head(image_embedding)  # (B, 80)
loss_cls = F.binary_cross_entropy_with_logits(cls_logits, label_vector)
loss_total = loss_contrastive + lambda_cls * loss_cls
```
 
### 4. Registry Entry
 
```yaml
B4:
  description: "B0+ + multi-label classification auxiliary loss (COCO 80 classes)"
  overrides:
    loss.cls_weight: 1.0
    loss.cls_num_classes: 80
```
 
## Evaluation
 
- Primary: R@1, R@5, R@10 (I2T and T2I), mAP@R on COCO 5K and Flickr30K test sets.
- Secondary: Log classification metrics (macro-averaged precision, recall, F1 over 80 classes) to monitor whether the head is actually learning.
- Compare against B0+ (same config minus classification head).
 
## Open Questions
 
- [ ] Does the current dataloader already load COCO metadata, or is it caption-only? → **Needs codebase investigation** (see prompt below).
- [ ] Should we also try class-weighted BCE (some COCO classes are very rare)?
- [ ] λ annealing schedule: cosine vs. step-wise vs. linear?
